client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,329 @@
+	}
+
+	/// Execute block, recording all spans and events belonging to `Self::targets`
+	/// and filter out events which do not have the keys starting with one of the
+	/// prefixes in `Self::storage_keys`.
+	pub fn trace_block(&self) -> Result<TraceBlockResponse, String> {
+		tracing::debug!(target: "state_tracing", "Tracing block: {}", self.block);
+		// Prepare the block
+		let id = BlockId::Hash(self.block);
+		let mut header = self.client.header(id)
+			.map_err(|e| format!("Invalid block Id: {:?}", e))?
+			.ok_or_else(|| "Header not found".to_string())?;
+		let extrinsics = self.client.block_body(&id)
+			.map_err(|e| format!("Invalid block Id: {:?}", e))?
+			.ok_or_else(|| "Extrinsics not found".to_string())?;
+		tracing::debug!(target: "state_tracing", "Found {} extrinsics", extrinsics.len());
+		let parent_hash = *header.parent_hash();
+		let parent_id = BlockId::Hash(parent_hash);
+		// Pop digest else RuntimePanic due to: 'Number of digest items must match that calculated.'
+		header.digest_mut().pop();
+		let block = Block::new(header, extrinsics);
+
+		let targets = if let Some(t) = &self.targets { t } else { DEFAULT_TARGETS };
+		let storage_keys = if let Some(s) = &self.storage_keys {
+			s
+		} else {
+			DEFAULT_STORAGE_KEYS
+		};
+		let spans = Mutex::new(HashMap::new());
+		let events = Mutex::new(Vec::new());
+		let block_subscriber = BlockSubscriber::new(targets, spans, events);
+		let dispatch = Dispatch::new(block_subscriber);
+
+		{
+			let dispatcher_span = tracing::debug_span!(
+				target: "state_tracing",
+				"execute_block",
+				extrinsics_len = block.extrinsics().len()
+			);
+			let _guard = dispatcher_span.enter();
+			if let Err(e) = dispatcher::with_default(&dispatch, || {
+				let span = tracing::info_span!(
+					target: TRACE_TARGET,
+					"trace_block",
+				);
+				let _enter = span.enter();
+				self.client.runtime_api().execute_block(&parent_id, block)
+			}) {
+				return Err(format!("Error executing block: {:?}", e));
+			}
+		}
+
+		let block_subscriber = dispatch.downcast_ref::<BlockSubscriber>()
+			.ok_or("Cannot downcast Dispatch to BlockSubscriber after tracing block")?;
+		let spans: Vec<_> = block_subscriber.spans
+			.lock()
+			.drain()
+			// Patch wasm identifiers
+			.filter_map(|(_, s)| patch_and_filter(SpanDatum::from(s), targets))
+			.collect();
+		let events: Vec<_> = block_subscriber.events
+			.lock()
+			.drain(..)
+			.filter(|e| event_key_filter(e, storage_keys))
+			.map(|s| s.into())
+			.collect();
+		tracing::debug!(target: "state_tracing", "Captured {} spans and {} events", spans.len(), events.len());
+
+		let block_trace = BlockTrace {
+			block_hash: block_id_as_string(id),
+			parent_hash: block_id_as_string(parent_id),
+			tracing_targets: targets.to_string(),
+			storage_keys: storage_keys.to_string(),
+			spans,
+			events,
+		};
+		let block_trace_size = serde_json::to_vec(&block_trace)
+			.map_err(|e| format!("Failed to serialize payload: {}", e))?
+			.len();
```

Comment:
This is a bit problematic, we serialize here and then once again before returning? I don't have a better idea atm but will try to come up with something.

I guess we mostly care about the `spans` and `events` members so perhaps we can keep a running count of them during execution? Or how about: we make a guesstimate of the average size of a `span` and `event` and then use that heuristic to calculate `approx_size_of_a_span * block_traces.spans.len()` (and same for events); I don't think we need to be very precise.

We could also move this check to where we return the JSON: serialize, check for length and if too big, return an error?

Reply:
yeah, that's a good idea I guess we could serialize the response to a String [here](https://github.com/paritytech/substrate/pull/7780/files#diff-d155229ffb6c782e34d8874b260cefb3b90d2b8549661b40c5b469f0fe125e0bR537) instead re-use it.

Reply:
> we make a guesstimate of the average size of a span and event and then use that heuristic to calculate approx_size_of_a_span * block_traces.spans.len() (and same for events); I don't think we need to be very precise.

If we want to go this route, I did some work to get estimates before [here](https://github.com/paritytech/substrate/pull/7780#issuecomment-819685718)

Reply:
Ok I just changed it to

```rust
		let aprox_payload_size = BASE_PAYLOAD + events.len() * AVG_EVENT + spans.len() * AVG_SPAN;
		let response = if aprox_payload_size > MAX_PAYLOAD {
				TraceBlockResponse::TraceError(TraceError { ... })
		} else {
			TraceBlockResponse::BlockTrace(BlockTrace { ... })
		};
```

Reply:
Ok i just changed it back to:

```rust
		let aprox_payload_size = BASE_PAYLOAD + events.len() * AVG_EVENT + spans.len() * AVG_SPAN;
		let response = if aprox_payload_size > MAX_PAYLOAD {
				TraceBlockResponse::TraceError(TraceError { ... })
		} else {
			TraceBlockResponse::BlockTrace(BlockTrace { ... })
		};
```

Reply:
> yeah, that's a good idea I guess we could serialize the response [here](https://github.com/paritytech/substrate/pull/7780/files#diff-d155229ffb6c782e34d8874b260cefb3b90d2b8549661b40c5b469f0fe125e0bR537) instead but not sure if it's possible to construct a `JsonRpcResponse` without double allocate anyway.

Yeah it looks a bit messy to get right. I say we punt on it and make do with the estimate Zeke implemented.